ext/bcmath/libbcmath/src/div.c

```diff
@@ -31,219 +31,440 @@
+	 * For example, the true quotient of 240 / 121 is 1.
+	 * However, if it is calculated using only the high-order digits (24 / 12), we would get 2.
+	 * We refer to this value of 2 as the temporary quotient.
+	 * We also define E to be error between the true quotient and the temporary quotient,
+	 * which in this case, is 1.
+	 * Also note that for example 2400000 / 120, there will be 5 divisions.
+	 *
+	 * Another example: Calculating 99900000000 / 10009999 with base 10000.
+	 * The actual quotient is 9980, but if calculate it using only the high-order digits:
+	 * 999 / 1000 = 0, Numbers that are not divisible are carried forward to the next division.
+	 * 9990000 / 1000 = 9990 (Here, the temporary quotient is 9990 and the error E is 10.)
+	 *
+	 * If calculate the temporary quotient using only one array of numerator and divisor, the error E can be larger than 1.
+	 * In other words, in the restoring division, the count of additions for restore increases significantly.
+	 *
+	 * Therefore, in order to keep the error within 1 and to limit the number of additions required for restoration to
+	 * at most one, adjust the number of high-order digits used to calculate the temporary quotient as follows.
+	 * - Adjust the number of digits of divisor used in the calculation to BC_VECTOR_SIZE + 1 digit. The missing digits are
+	 *   filled in from the next array element.
+	 * - Add digits to numerator in the same way as the number of digits adjusted by divisor.
+	 *
+	 * e.g.
+	 * numerator = 123456780000
+	 * divisor = 123456789
+	 * base = 10000
+	 * numerator_arr = [0, 5678, 1234]
+	 * divisor_arr = [6789, 2345, 1]
+	 * numerator_top = 1234
+	 * divisor_top = 1
+	 * divisor_top_tmp = 12345 (+4 digits)
+	 * numerator_top_tmp = 12345678 (+4 digits)
+	 * tmp_quot = numerator_top_tmp / divisor_top_tmp = 1000
+	 * tmp_rem = -9000
+	 * tmp_quot is too large, so tmp_quot--, tmp_rem += divisor (restoring)
+	 * quot = 999
+	 * rem = 123447789
+	 *
+	 * Details:
+	 * Suppose that when calculating the temporary quotient, only the high-order elements of the BC_VECTOR array are used.
+	 * At this time, numerator and divisor can be considered to be decomposed as follows. (Here, B = 10^b, any b ∈ ℕ, any k ∈ ℕ)
+	 * numerator = numerator_high * B^k + numerator_low
+	 * divisor = divisor_high * B^k + divisor_low
+	 *
+	 * The error E can be expressed by the following formula.
+	 *
+	 * E = (numerator_high * B^k) / (divisor_high * B^k) - (numerator_high * B^k + numerator_low) / (divisor_high * B^k + divisor_low)
+	 * <=> E = numerator_high / divisor_high - (numerator_high * B^k + numerator_low) / (divisor_high * B^k + divisor_low)
+	 * <=> E = (numerator_high * (divisor_high * B^k + divisor_low) - (numerator_high * B^k + numerator_low) * divisor_high) / (divisor_high * (divisor_high * B^k + divisor_low))
+	 * <=> E = (numerator_high * divisor_low - divisor_high * numerator_low) / (divisor_high * (divisor_high * B^k + divisor_low))
```

Comment:
I think one thing you need to justify is why `0 ≤ E`.

This looks correct, but I am not fully convinced.

The simplification is correct as I just went through it with pen and paper

Reply:
Since this is a Restoring division, I don't think the error will ever be negative.

https://en.wikipedia.org/wiki/Division_algorithm#Restoring_division

Reply:
Right, I didn't see that the error was just a proxy for the restoration number.

Reply:
Do you think stating that it's a restoring division, and listing this URL in the comments, would clarify things?

Reply:
Possibly just before the function parameter list yes

Reply:
Added
https://github.com/php/php-src/pull/14660/commits/3cca73df462db70809dbe611ba977fddf5c9fabe